# Taller 3: Exploratory Data Analysis (EDA) y Data Wrangling

## Introducción

Instacart es una plataforma de entrega de comestibles donde los clientes pueden hacer un pedido y recibirlo, de manera similar a como funcionan Uber Eats y Door Dash.
El conjunto de datos que se te ha proporcionado ha sido modificado del original. Hemos reducido el tamaño del conjunto para que tus cálculos se ejecuten más rápido y hemos introducido deliberadamente valores ausentes (missing values) y duplicados. Sin embargo, hemos conservado las distribuciones poblacionales de los datos originales.

**Tu misión profesional:** Como Científico de Datos, aplicarás metodológicamente las 4 dimensiones de la calidad de los datos para auditar tu dataset. Luego, realizarás un análisis exploratorio empleando medidas de tendencia central, variabilidad, diagramas de dispersión, y responderás preguntas de negocio tras construir una tabla desnormalizada (One Big Table - OBT).

> **Aviso Importante:** Para cada paso, escribe introducciones donde expongas qué harás, bloques intermedios justificando el porqué de tus decisiones y tu limpieza estadística, y cierra siempre con una conclusión de lo hallado.

## Diccionario de Datos

Hay cinco tablas en el conjunto de datos que deberás usar en conjunto:

- `instacart_orders.csv`: cada fila es un pedido en la app de Instacart
    - `'order_id'`: ID que identifica de manera única cada pedido
    - `'user_id'`: ID que identifica de manera única a cada cliente
    - `'order_number'`: el número de veces que este cliente ha hecho un pedido
    - `'order_dow'`: día de la semana en que se hizo el pedido (0 es domingo)
    - `'order_hour_of_day'`: hora del día en que se hizo el pedido
    - `'days_since_prior_order'`: número de días desde que este cliente hizo su último pedido
- `products.csv`: cada fila corresponde a un producto
    - `'product_id'`: ID único del producto
    - `'product_name'`: nombre del producto
    - `'aisle_id'`: ID único de la categoría del pasillo
    - `'department_id'`: ID único de la categoría del departamento
- `order_products.csv`: cada fila corresponde a un artículo en un pedido
    - `'order_id'`: ID único del pedido
    - `'product_id'`: ID único del producto
    - `'add_to_cart_order'`: el orden secuencial en el que se añadió cada artículo al carrito
    - `'reordered'`: 0 si el cliente no había pedido este producto antes, 1 si ya lo había hecho
- `aisles.csv`
    - `'aisle_id'`: ID único del pasillo
    - `'aisle'`: nombre del pasillo
- `departments.csv`
    - `'department_id'`: ID único del departamento
    - `'department'`: nombre del departamento

# Paso 1: Exploración Inicial

Lee los archivos ubicados en el directorio `/datos/` (`instacart_orders.csv`, `products.csv`, `aisles.csv`, `departments.csv` y `order_products.csv`) fijándote que algunos están separados por `;` u otros delimitadores. 

Tras leerlos, ejecuta funciones de inspección básica (`head()`, `info()`) y comprueba una muestra aleatoria (Muestreo Aleatorio Simple usando `sample()`) de cada tabla para evitar el *sesgo de posición*. Verifica si los tipos computacionales coinciden con su rol estadístico esperado.

In [1]:
# importar librerías
import pandas as pd

# Instacart_orders

In [2]:
# lectura de dataset instacart_orders.csv desde la carpeta datos usando pandas
orders = pd.read_csv('datos/instacart_orders.csv', sep=';')

# Inspección de las primeras filas del dataset
orders.head()

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,1515936,183418,11,6,13,30.0
1,1690866,163593,5,5,12,9.0
2,1454967,39980,4,5,19,2.0
3,1768857,82516,56,0,20,10.0
4,3007858,196724,2,4,12,17.0


In [3]:
# Información de las características del dataset orders
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478967 entries, 0 to 478966
Data columns (total 6 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   order_id                478967 non-null  int64  
 1   user_id                 478967 non-null  int64  
 2   order_number            478967 non-null  int64  
 3   order_dow               478967 non-null  int64  
 4   order_hour_of_day       478967 non-null  int64  
 5   days_since_prior_order  450148 non-null  float64
dtypes: float64(1), int64(5)
memory usage: 21.9 MB


In [4]:
# Muestreo aleatorio de filas del dataset orders
orders.sample(10)

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
367462,1828322,34075,4,5,14,22.0
217368,1971252,203341,58,3,15,8.0
471027,3105701,148848,8,3,15,16.0
427520,107793,968,57,3,12,8.0
159910,2598518,205730,10,6,21,3.0
438208,2060740,138395,1,1,14,NaN
281002,2521618,140395,83,6,8,1.0
336302,2506060,48215,11,0,11,6.0
135785,2689123,163753,53,2,17,2.0
258546,2204339,143539,4,1,8,14.0


In [5]:
# Estadísticas descriptivas del dataset orders
orders.describe()

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
count,4.789670e+05,478967.000000,478967.000000,478967.000000,478967.000000,450148.000000
mean,1.709947e+06,102966.537475,17.157808,2.775058,13.446676,11.101751
std,9.869480e+05,59525.328991,17.704688,2.045869,4.224986,9.189975
min,4.000000e+00,2.000000,1.000000,0.000000,0.000000,0.000000
25%,8.564685e+05,51414.000000,5.000000,1.000000,10.000000,4.000000
50%,1.709522e+06,102633.000000,11.000000,3.000000,13.000000,7.000000
75%,2.562990e+06,154419.000000,23.000000,5.000000,16.000000,15.000000
max,3.421079e+06,206209.000000,100.000000,6.000000,23.000000,30.000000


### Escribe tus hallazgos iniciales aquí:
(Escribe textualmente qué dimensiones, tipos de datos ves incoherentes y qué opinas estadísticamente de la primera radiografía de los datos).

- Existen 6 columnas en el dataset y 478967 filas.
- Los tipos de datos se ven coherentes, sin embargo, days_since_prior_order se muestra como datos flotantes, a pesar de tener datos numéricos de 0 a 30 sin decimales.
- El dataset dispone de valores faltantes en days_since_prior_order, ya que existen 450148 registro de un total de 478967 filas.
- Considerando la estadística de los datos, se puede observar que el dataset es una muestra, ya que la numeración máxima de la orden es 3 421 079 y el mínimo 4, lo cual no coincide con los más de 400000 registros.

# Products.csv

In [6]:
# lectura de dataset products.csv desde la carpeta datos usando pandas
products = pd.read_csv('datos/products.csv', sep=';')

# Inspección de las primeras filas del dataset
products.head(10)

,product_id,product_name,aisle_id,department_id
0,1,Chocolate Sandwich Cookies,61,19
1,2,All-Seasons Salt,104,13
2,3,Robust Golden Unsweetened Oolong Tea,94,7
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1
4,5,Green Chile Anytime Sauce,5,13
5,6,Dry Nose Oil,11,11
6,7,Pure Coconut Water With Orange,98,7
7,8,Cut Russet Potatoes Steam N' Mash,116,1
8,9,Light Strawberry Blueberry Yogurt,120,16
9,10,Sparkling Orange Juice & Prickly Pear Beverage,115,7


In [7]:
# Información de las características del dataset products
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49694 entries, 0 to 49693
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49694 non-null  int64 
 1   product_name   48436 non-null  object
 2   aisle_id       49694 non-null  int64 
 3   department_id  49694 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB


In [8]:
# Muestreo aleatorio de filas del dataset products
products.sample(5)

,product_id,product_name,aisle_id,department_id
47638,47639,Blueberry Protein Shake,98,7
21345,21346,Zinc 50 mg,47,11
10788,10789,Pure Aloe Whole Leaf Juice with Lemon,77,7
350,351,Go Fresh Nutrium Moisture Body Wash Blue Fig &...,127,11
19944,19945,G Series Perform Orange Flavor Sports Drink,64,7


In [9]:
# Estadísticas descriptivas del dataset products
products.describe()

,product_id,aisle_id,department_id
count,49694.000000,49694.000000,49694.000000
mean,24847.500000,67.773192,11.728478
std,14345.566475,38.317265,5.850314
min,1.000000,1.000000,1.000000
25%,12424.250000,35.000000,7.000000
50%,24847.500000,69.000000,13.000000
75%,37270.750000,100.000000,17.000000
max,49694.000000,134.000000,21.000000


### Escribe tus hallazgos iniciales aquí:
(Escribe textualmente qué dimensiones, tipos de datos ves incoherentes y qué opinas estadísticamente de la primera radiografía de los datos).

- Existen 4 columnas en el dataset y 49694 filas.
- Los tipos de datos son correctos.
- El dataset dispone de valores faltantes en product_name, ya que existen 48436 registros de un total de 49694 filas.
- Considerando la estadística de los datos, se puede observar que existen 21 departamentos y 134 islas, además de 49694 productos. Esto nos indica que existe la población total de productos.

# Aisles.csv

In [10]:
# lectura de dataset aisles.csv desde la carpeta datos usando pandas
aisles = pd.read_csv('datos/aisles.csv', sep=';')

# Inspección de las primeras filas del dataset
aisles.head()

,aisle_id,aisle
0,1,prepared soups salads
1,2,specialty cheeses
2,3,energy granola bars
3,4,instant foods
4,5,marinades meat preparation


In [11]:
# Información de las características del dataset aisles
aisles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134 entries, 0 to 133
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   aisle_id  134 non-null    int64 
 1   aisle     134 non-null    object
dtypes: int64(1), object(1)
memory usage: 2.2+ KB


In [12]:
# Muestreo aleatorio de filas del dataset aisles
aisles.sample(10)

,aisle_id,aisle
101,102,baby bath body care
21,22,hair care
132,133,muscles joints pain relief
33,34,frozen meat seafood
45,46,mint gum
53,54,paper goods
6,7,packaged meat
102,103,ice cream toppings
92,93,breakfast bakery
57,58,frozen breads doughs


In [13]:
# Estadísticas descriptivas del dataset aisles
aisles.describe()

,aisle_id
count,134.000000
mean,67.500000
std,38.826537
min,1.000000
25%,34.250000
50%,67.500000
75%,100.750000
max,134.000000


### Escribe tus hallazgos iniciales aquí:
(Escribe textualmente qué dimensiones, tipos de datos ves incoherentes y qué opinas estadísticamente de la primera radiografía de los datos).

- Existen 2 columnas en el dataset y 134 filas.
- Los tipos de datos son coherentes.
- No existen valores faltantes.
- Considerando la estadística de los datos, se puede observar que existe un orden en la numeración de las islas. 
- Adicionalmente, la numeración y total de filas coincide, por lo que el dataset corresponde al total de los datos, por lo que corresponde a la población. 

# Departments.csv

In [14]:
# lectura de dataset departments.csv desde la carpeta datos usando pandas
departments = pd.read_csv('datos/departments.csv', sep=';')

# Inspección de las primeras filas del dataset
departments.head()

,department_id,department
0,1,frozen
1,2,other
2,3,bakery
3,4,produce
4,5,alcohol


In [15]:
# Información de las características del dataset departments
departments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   department_id  21 non-null     int64 
 1   department     21 non-null     object
dtypes: int64(1), object(1)
memory usage: 468.0+ bytes


In [16]:
# Muestreo aleatorio de filas del dataset departments
departments.sample(5)

,department_id,department
9,10,bulk
7,8,pets
16,17,household
20,21,missing
4,5,alcohol


In [17]:
# Estadísticas descriptivas del dataset departments
departments.describe()

,department_id
count,21.000000
mean,11.000000
std,6.204837
min,1.000000
25%,6.000000
50%,11.000000
75%,16.000000
max,21.000000


### Escribe tus hallazgos iniciales aquí:
(Escribe textualmente qué dimensiones, tipos de datos ves incoherentes y qué opinas estadísticamente de la primera radiografía de los datos).

- Existen 2 columnas en el dataset y 21 filas.
- Los tipos de datos son coherentes.
- El dataset no dispone de valores faltantes.
- Considerando la estadística de los datos, se puede observar que existen 21 departamentos. Esto nos indica que existe la población total de departamentos.

# Order_products.csv

In [18]:
# lectura de dataset order_products.csv desde la carpeta datos usando pandas
order_products = pd.read_csv('datos/order_products.csv', sep=';')

# Inspección de las primeras filas del dataset
order_products.head()

,order_id,product_id,add_to_cart_order,reordered
0,2141543,11440,17.0,0
1,567889,1560,1.0,1
2,2261212,26683,1.0,1
3,491251,8670,35.0,1
4,2571142,1940,5.0,1


In [19]:
# Información de las características del dataset order_products
order_products.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4545007 entries, 0 to 4545006
Data columns (total 4 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   order_id           4545007 non-null  int64  
 1   product_id         4545007 non-null  int64  
 2   add_to_cart_order  4544171 non-null  float64
 3   reordered          4545007 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 138.7 MB


In [20]:
# Muestreo aleatorio de filas del dataset order_products
order_products.sample(10)

,order_id,product_id,add_to_cart_order,reordered
3246057,2011388,43394,2.0,1
40032,2358864,1999,35.0,0
2551385,927835,2086,4.0,1
4174478,2313965,27086,5.0,1
4323734,3211842,14233,4.0,1
3247222,3196334,13176,17.0,1
3924186,3385316,24382,5.0,0
3095811,2907123,27521,13.0,1
2787187,272687,41328,2.0,0
1125958,3323340,21526,12.0,1


In [21]:
# Estadísticas descriptivas del dataset order_products
order_products.describe()

,order_id,product_id,add_to_cart_order,reordered
count,4.545007e+06,4.545007e+06,4.544171e+06,4.545007e+06
mean,1.711166e+06,2.558084e+04,8.351669e+00,5.905025e-01
std,9.850955e+05,1.409552e+04,7.080368e+00,4.917411e-01
min,4.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00
25%,8.608170e+05,1.353500e+04,3.000000e+00,0.000000e+00
50%,1.711530e+06,2.527200e+04,6.000000e+00,1.000000e+00
75%,2.561311e+06,3.793500e+04,1.100000e+01,1.000000e+00
max,3.421079e+06,4.969400e+04,6.400000e+01,1.000000e+00


### Escribe tus hallazgos iniciales aquí:
(Escribe textualmente qué dimensiones, tipos de datos ves incoherentes y qué opinas estadísticamente de la primera radiografía de los datos).

- Existen 4 columnas en el dataset y 4545007 filas.
- add_to_cart_order está especificado como float, sin embargo es un orden, por lo que no pueden existir número decimales.
- El dataset dispone de valores faltantes en add_to_cart_order, ya que existen 4544171 registros de un total de 4545007 filas.

# Paso 2: Auditoría de Calidad de Datos (4 Dimensiones)

Basado en la teoría vista, somete a tus 5 tablas al escrutinio en las cuatro dimensiones.

## 2.1 Precisión: Identificación y limpieza de duplicados explícitos e implícitos
Encuentra y elimina duplicados explícitos en la tabla de `orders`. Nota si se agrupan por algún motivo (¿Son de un día y hora en particular?). Asimismo, busca duplicados por ID de producto (tricky duplicates) o nombres duplicados con problemas de mayúsculas y minúsculas.

* Duplicados Instacart_orders 

In [22]:
# Código para encontrar duplicados en la tabla orders
orders.duplicated().sum()

np.int64(15)

In [23]:
# visualizarción de filas duplicadas en la tabla orders
orders[orders.duplicated()]

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
145574,794638,50898,24,3,2,2.0
223105,2160484,107525,16,3,2,30.0
230807,1918001,188546,14,3,2,16.0
266232,1782114,106752,1,3,2,NaN
273805,1112182,202304,84,3,2,6.0
284038,2845099,31189,11,3,2,7.0
311713,1021560,53767,3,3,2,9.0
321100,408114,68324,4,3,2,18.0
323900,1919531,191501,32,3,2,7.0
345917,2232988,82565,1,3,2,NaN


In [24]:
# Código para contar registros de order_dow=3 y order_hour_of_day=2 en la tabla orders
orders[(orders['order_dow'] == 3) & (orders['order_hour_of_day'] == 2)].shape[0]

121

In [25]:
# eliminación definitiva de filas duplicadas en la tabla orders
orders.drop_duplicates(inplace=True)

In [26]:
#verificación de duplicados de numero de orden en la tabla orders
orders.duplicated(subset=['order_id']).sum()

np.int64(0)

* Duplicados Products

In [27]:
# búsqueda de duplicados explícitos en la tabla products
orders.duplicated().sum()

np.int64(0)

In [28]:
# Análisis de tricky duplicates en la tabla products
# Mismo products_id pero diferente product_name
products[products['product_id'].duplicated()]

,product_id,product_name,aisle_id,department_id


In [29]:
# verificar nombres duplicados
# cambiar nombres a minúscula para evitar problemas de mayúscula/minúscula
products['product_name'] = products['product_name'].str.lower()

In [30]:
# Con el mismo nombre pero diferente id
products[products.duplicated(subset=['product_name'], keep=False)].sort_values('product_name')

,product_id,product_name,aisle_id,department_id
23339,23340,18-in-1 hemp peppermint pure-castile soap,25,11
31844,31845,18-in-1 hemp peppermint pure-castile soap,25,11
19941,19942,aged balsamic vinegar of modena,19,13
13152,13153,aged balsamic vinegar of modena,19,13
24830,24831,albacore solid white tuna in water,95,15
...,...,...,...,...
49552,49553,NaN,100,21
49574,49575,NaN,100,21
49640,49641,NaN,100,21
49663,49664,NaN,100,21


In [31]:
# Nombres repetidos que no corresponden a NaN
products[products.duplicated(subset=['product_name'], keep=False) & products['product_name'].notna()].sort_values('product_name')

,product_id,product_name,aisle_id,department_id
23339,23340,18-in-1 hemp peppermint pure-castile soap,25,11
31844,31845,18-in-1 hemp peppermint pure-castile soap,25,11
19941,19942,aged balsamic vinegar of modena,19,13
13152,13153,aged balsamic vinegar of modena,19,13
24830,24831,albacore solid white tuna in water,95,15
...,...,...,...,...
46873,46874,vitamin d3 5000 iu,47,11
21366,21367,wild sardines in spring water,95,15
40093,40094,wild sardines in spring water,95,15
1826,1827,yams cut sweet potatoes in syrup,81,15


* Duplicados order_products

In [32]:
# Búsqueda de duplicados explícitos en la tabla order_products
order_products[order_products.duplicated()]

,order_id,product_id,add_to_cart_order,reordered


* Duplicados Aisles

In [33]:
# Verificación de duplicados en la tabla aisles
aisles[aisles['aisle_id'].duplicated()]

,aisle_id,aisle


In [34]:
# Análisis de tricky duplicates en la tabla aisles
# Mismo aisle_id pero diferente aisle
aisles[aisles.duplicated(subset=['aisle_id'])]

,aisle_id,aisle


In [35]:
# mismo nombre de aisle pero diferente id
# cambio de nombres a minúscula para evitar problemas de mayúscula/minúscula
aisles['aisle'] = aisles['aisle'].str.lower()
aisles[aisles.duplicated(subset=['aisle'])]

,aisle_id,aisle


* Duplicados Departments

In [36]:
# Verificación de duplicados en la tabla departments

departments[departments['department_id'].duplicated()]

,department_id,department


In [37]:
# Análisis de tricky duplicates en la tabla departments
# Mismo department_id pero diferente department
departments[departments.duplicated(subset=['department_id'])]

,department_id,department


In [38]:
# mismo nombre de departments pero diferente id
# cambio de nombres a minúscula para evitar problemas de mayúscula/minúscula
departments['department'] = departments['department'].str.lower()
departments[departments.duplicated(subset=['department'])]

,department_id,department


### Conclusiones sobre la Precisión:
Explica por qué eliminaste (o consolidaste) la información y qué nos enseña sobre la ingesta del sistema.

* Instacart_Orders:
   - Se encontraron duplicados explícitos, correspondientes a 15 filas, las cuales fueron eliminadas. No se encontraron tricky duplicates.

* Products:
   - No se encontraron duplicados explícitos.
   - Existen 207 filas duplicadas en el nombre de producto, sin embargo, se mantendrán, ya que difieren en product_id.
   - Existen más de 1000 productos con nombre faltante que corresponden al departamento 21 e isla 100.
   - La isla 100 tiene nombre faltante al igual que el departamento 21

* Order_products:
   - No se encontraron duplicados explícitos.
   - No se realizó la búsqueda de tricky duplicates, ya que por la naturaleza de la tabla, existirán duplicados en todas las columnas.

* Aisles:
   - No se encontraron duplicados explícitos ni tricky.

* Departments:
   - No se encontraron duplicados explícitos ni tricky.

## 2.2 Completitud: Tratamiento de Valores Ausentes
Navega por las tablas en busca de NaNs. Es crítico que analices si el patrón de valores ausentes se trata de MCAR (Missing Completely at Random), MAR o MNAR.

1. Revisa nombres de productos vacíos.
2. Revisa NaNs en `days_since_prior_order`. (¿Son clientes haciendo su primera orden?)
3. Revisa `add_to_cart_order` en `order_products`. (¿Qué pasa después de 64 artículos?)

Decide si debes imputar por tendencia central, usar ML, imputar un valor indicador ficticio (e.g. -1 o 'Desconocido') o eliminar las filas. ¡Aplica tus soluciones!

Tratamiento de valores ausentes en Tabla Instacart Orders

In [39]:
# Existen valores NAN en las siguientes tablas: orders, products y order_products
# conteo de valores NAN en la tabla orders
#visualización de valores faltantes en la tabla orders
orders[orders.isna().any(axis=1)]

,order_id,user_id,order_number,order_dow,order_hour_of_day,days_since_prior_order
28,133707,182261,1,3,10,NaN
96,787445,25685,1,6,18,NaN
100,294410,111449,1,0,19,NaN
103,2869915,123958,1,4,16,NaN
104,2521921,42286,1,3,18,NaN
...,...,...,...,...,...,...
478895,2589657,205028,1,0,16,NaN
478896,2222353,141211,1,2,13,NaN
478922,2272807,204154,1,1,15,NaN
478926,2499542,68810,1,4,19,NaN


In [40]:
# valor de order_number para days_since_prior_order con valor NaN
orders[orders['days_since_prior_order'].isna()].order_number.nunique()

1

In [41]:
# Llenar valores NaN en days_since_prior_order con -1 para indicar que no hay un pedido previo
orders['days_since_prior_order'] = orders['days_since_prior_order'].fillna(-1)

Tratamiento de valores ausentes en Tabla Products

In [42]:
# Verificar nulos en products
products[products.isna().any(axis=1)]

,product_id,product_name,aisle_id,department_id
37,38,NaN,100,21
71,72,NaN,100,21
109,110,NaN,100,21
296,297,NaN,100,21
416,417,NaN,100,21
...,...,...,...,...
49552,49553,NaN,100,21
49574,49575,NaN,100,21
49640,49641,NaN,100,21
49663,49664,NaN,100,21


In [43]:
#verificar si todos los productos con aisle_id=100 corresponden al mismo product_name
products[products['aisle_id'] == 100]['product_name'].unique()

array([nan], dtype=object)

In [44]:
# Verificar si todos los productos con department_id=21 corresponden al mismo product_name
products[products['department_id'] == 21]['product_name'].unique()

array([nan], dtype=object)

In [45]:
# cambio de nombres nan por Desconocido
products['product_name'] = products['product_name'].fillna('Desconocido')

Tratamiento de valores ausentes en Tabla Order_products

In [54]:
# Análisis de valores faltantes en la tabla order_products
order_products.isna().sum()

order_id               0
product_id             0
add_to_cart_order    836
reordered              0
dtype: int64

In [61]:
# Visualización de filas con valores faltantes en la tabla order_products
order_products[order_products['add_to_cart_order'].isna()]

,order_id,product_id,add_to_cart_order,reordered
737,2449164,5068,NaN,0
9926,1968313,43867,NaN,0
14394,2926893,11688,NaN,0
16418,1717990,4142,NaN,0
30114,1959075,42828,NaN,1
...,...,...,...,...
4505662,1800005,7411,NaN,0
4511400,1633337,260,NaN,0
4517562,404157,9517,NaN,0
4534112,1673227,17835,NaN,0


In [75]:
# Estas son las ordenes que tienen productos sin posicion del carrito de compras
ordenes_faltantes = order_products[order_products['add_to_cart_order'].isna()]['order_id']
ordenes_faltantes.value_counts()

order_id
61355      63
3308010    51
2136777    44
171934     40
1959075    34
           ..
1598369     1
9310        1
2729254     1
2621907     1
2170451     1
Name: count, Length: 70, dtype: int64

In [76]:
ordenes_faltantes = ordenes_faltantes.unique()

In [78]:
full_detalle_ordenes_con_posicion_na = order_products[order_products['order_id'].isin(ordenes_faltantes)]
full_detalle_ordenes_con_posicion_na['order_id'].value_counts()

order_id
61355      127
3308010    115
2136777    108
171934     104
1959075     98
          ... 
888470      65
1598369     65
747668      65
1677118     65
2170451     65
Name: count, Length: 70, dtype: int64

In [79]:
order_products['add_to_cart_order'] = order_products['add_to_cart_order'].fillna(-1).astype('int')

### Conclusiones sobre la Completitud:
Argumenta qué mecanismo de ausencia (Rubin) encontraste en cada escenario y por qué tus estrategias de imputación no generaron sesgos severos.

* Instacart_Orders:
   - days_since_prior_order presentó valores ausentes del tipo MNAR — Missing Not at Random, ya que los valores ausentes corresponden a la primera compra. Para este caso los valores  fueron llenados con -1 para no afectar a los datos actualemnte en el sistema.

* Products:
   - product_name presentó valores ausentes del tipo MNAR — Missing Not at Random, ya que todos los nombres faltantes corresponden a la isla 100 y departamento 21. En este caso, todos los nombres fueron identificados "Desconocido".

* Order_products:
   -    add_to_cart_order presentó valores ausentes del tipo MNAR — Missing Not at Random, ya que todos los valores faltantes corresponden a compras superiores a 64 items. En este caso, todos los valores fueron reemplazados por -1.

## 2.3 Sensibilidad: Tratamiento de Outliers (Valores Atípicos)
Centra tu atención en la tabla de órdenes, específicamente en `days_since_prior_order` o conteos agrupados. Usando gráficos estadísticos (Boxplots) y computando el Rango Intercuartílico (IQR / Regla de Tukey) o Z-Scores, encuentra y reporta valores atípicos. ¿Existen usuarios que compran muy fuera del rango promedio de frecuencia?

Tabla Instacart_orders

In [85]:
# Boxplot de days_since_prior_order para verificar usuarios que compran fuer del rango promedio de frecuencia de compra
import matplotlib.pyplot as plt

ModuleNotFoundError: No module named 'matplotlib'

### Conclusiones sobre la Sensibilidad:
Decide y justifica la estrategia que usaste para estos casos (eliminar, mantener como outlier real o winsorizar).

# Paso 3: Data Wrangling & Construcción de One Big Table (OBT)

Para poder responder preguntas de negocio de manera eficiente y aplicando correlaciones y cruces, un científico de datos típicamente convierte o "aplana" las tablas normalizadas operacionales en **One Big Table (OBT)**. La OBT agrupa (hace merges) de las distintas dimensiones transaccionales alrededor del concepto central (en este caso el *Articulo pedido por Orden*), lo que nos da una vista panorámica (desnormalización estructural) para el análisis.

**Tu tarea:** Une las cinco tablas previamente purgadas formando un solo DataFrame analítico. Te recomendamos ir uniendo ordenes y el detalle, luego producto, pasillo y finalmente departamento.

# Paso 4: Análisis Multivariado y de Negocio

Con tu conjunto de datos ahora limpio (*trusted/gold data*) mediante la auditoría y unificado en tu OBT, usa tus conocimientos de agregación y visualización estadística para responder las siguientes interrogantes. Usa histogramas, KDE y/o gráficos de dispersión donde la distribución lo merezca.

## [A] Preguntas Esenciales

**A1. La hora y el día: Verificación del dominio:**
Verifica con código que `order_hour_of_day` y `order_dow` tienen distribuciones lógicas basándonos en tu conocimiento del mundo real. Construye histogramas para ver qué picos de horas concentran compras. Aplica gráficos de barras para evaluar diferencias de días y horas (comparemos, por ejemplo, distribuciones de la demanda los Miércoles vs. Sábados). Verifica si existe una diferencia de horas por dia.

**A2. Distribuciones temporales de recompra:**
¿Cuánto tiempo transcurre estadísticamente para que alguien vuelva a realizar otra orden? Grafica este tiempo de espera y concluye sobre la concentración de los datos (¿asimetría?).

### Tira aquí tus hallazgos estadísticos para el grupo A:
(Debes hablar de cómo se distribuye la demanda, tendencias centrales encontradas y asimetrías de cola).

## [B] Profundización (Segmentación Categórica)

**B1. Retención y recurrencia (Número de órdenes por usuario):**
Aislando o agrupando a nivel de granularidad de cliente, describe la variabilidad y distribución de compras que hacen. ¿Existen colas largas de clientes extra-leales?

**B2. Productos Top: El principio de Pareto:**
Genera un top 20 de los productos más solicitados globalmente. Para cada producto reporta su ratio de recompra. ¿Algunos productos tienen una correlación fuerte entre ser de un 'pasillo' y volverse recompras aseguradas?

### Conclusiones sobre el perfil del carrito [Grupo B]:

## [C] Patrones de Causalidad y Exposición (Hard)

**C1. El tamaño de la canasta comercial:**
¿Cuántos artículos en promedio compran las personas estadísticamente? Aplica gráficos sobre esta proporción y analiza la curtosis (grado de acumulación) de esta distribución del tamaño del pedido.

**C2. ¿Son los primeros productos un síntoma de fidelidad? (Spearman o Pearson):**
Para los 20 objetos que con mayor frecuencia las personas agregan como posición #1 al carrito, elabora una lógica que calcule una matriz de correlación (en variables generadas o agrupadas) o genera gráficos de dispersión (Scatterplot). Compara si, a nivel de producto, ser posicionado primero tiene relación lineal con ser un producto de constante recompra global (el campo 'reordered').

**C3. Para cada producto, ¿cual es la proporcion de re-compras?**

**C4. Para cada usuario, ¿cual es la proporcion de re-compras de los productos comprados?**

**C5. ¿Cuales son los top 20 productos que los clientes ponen primero en sus carritos?**

### Tus conclusiones avanzadas [Grupo C]:
(Enuncia las posibles relaciones matemáticas encontradas o variables dispersas)

# Conclusión General Ejecutiva
Resume en un párrafo las principales fortalezas, falencias detectadas en la gobernanza inicial de este dataset, y los descubrimientos de negocio clave que proporcionarías a la directiva de Instacart.